pf/internal/check/checks.go

```diff
@@ -12,20 +12,44 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package tfgen
+package check
 
 import (
+	"errors"
 	"fmt"
 
+	"github.com/pulumi/pulumi/sdk/v3/go/common/diag"
+
+	"github.com/pulumi/pulumi-terraform-bridge/pf/internal/muxer"
 	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge"
 	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim"
-	"github.com/pulumi/pulumi/sdk/v3/go/common/diag"
 )
 
-func checkIDProperties(sink diag.Sink, info tfbridge.ProviderInfo) error {
+// Validate that info is valid as either a PF provider or a PF & SDK based provider.
+//
+// This function should be called in the generate step, but before schema generation (so
+// as to error as soon as possible).
+func Provider(sink diag.Sink, info tfbridge.ProviderInfo) error {
+	isPFResource := func(string) bool { return true }
+	isPFDataSource := func(string) bool { return true }
+	if p, ok := info.P.(*muxer.ProviderShim); ok {
+		isPFResource = p.ResourceIsPF
+		isPFDataSource = p.DataSourceIsPF
+	}
+
+	return errors.Join(
+		checkIDProperties(sink, info, isPFResource),
+		notSupported(sink, info, isPFResource, isPFDataSource),
+	)
+}
+
+func checkIDProperties(sink diag.Sink, info tfbridge.ProviderInfo, doCheck func(tfToken string) bool) error {
 	errors := 0
 
 	info.P.ResourcesMap().Range(func(rname string, resource shim.Resource) bool {
+		if !doCheck(rname) {
```

Comment:
I think this would be more clear if it were still called `isPfResource`. Ideally we would only call `checkIDProperties` at all if the resource in question is a PF resource, but since that's not how the (muxed) resources map is set up, I think a name that says "if this _isn't_ a PF resource, we always return true" is more clear. The decision to do the check is inherent to the PF-y-ness.

Reply:
I added a name change and a comment in 83fd9b4eb5b338c66ba9699b2cbdb12a31245315.